### PR TITLE
Support for AMDGCN/ROCm.

### PR DESCRIPTION
**Pull request by stuartarchibald**
This patch adds support targetting AMD discrete GPUs as a compute
device. The work in the patch was carried out in relay by @sklam
and @stuartarchibald.

### CODE REVIEW

File: numba/roc/hsadrv/driver.py [hunk 1]
```diff
@@ -0,0 +1,1586 @@
...
+            expire = timeout * hsa.timestamp_frequency * mhz
+
+        # XXX: use active wait instead of blocked seem to avoid hang in docker
+        hsa.hsa_signal_wait_acquire(self._id, enums.HSA_SIGNAL_CONDITION_NE,
+                                    one, expire,
+                                    enums.HSA_WAIT_STATE_ACTIVE)
+        return self.load_relaxed() != one
+
+
+class BrigModule(object):
```
Comment (stuartarchibald) [inline numba/roc/hsadrv/driver.py]: Not convinced this is used or needed?

File: numba/roc/hsadrv/driver.py [hunk 2]
```diff
@@ -0,0 +1,1586 @@
...
+        signals = (0, None, completion_signal)
+
+    hsa.hsa_amd_memory_async_copy(dst, dst_ctx._agent._id,
+                                  src, src_ctx._agent._id,
+                                  size, *signals)
+
+    stream._add_signal(completion_signal)
+
+
+# Known Hardware
```
Comment (stuartarchibald) [inline numba/roc/hsadrv/driver.py]: Don't think this is needed.

File: numba/roc/hsadrv/driver.py [hunk 3]
```diff
@@ -0,0 +1,1586 @@
...
+known_dgpus = frozenset([b'Fiji'])
+known_apus = frozenset([b'Spectre'])
+known_cpus = frozenset([b'Kaveri'])
+
+def apu_present():
```
Comment (stuartarchibald) [inline numba/roc/hsadrv/driver.py]: Remove this, dGPU support only?

File: numba/roc/tests/hsapy/run_far_branch.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        sin_delta, cos_delta = math.sin(delta), math.cos(delta)
+
+        numerator = math.sqrt((cos_lat1 * sin_delta)**2 +
+                                (cos_lat1 * sin_lat2 - sin_lat1 * cos_lat2 * cos_delta)**2)
+        denominator = sin_lat1 * sin_lat2 + cos_lat1 * cos_lat2 * cos_delta
+        return R_EARTH * math.atan2(numerator, denominator)
+
+    arr = np.random.random(10).astype(np.float32)
+
+    print(gpu_great_circle_distance(arr, arr, arr, arr))
```
Comment (stuartarchibald) [inline numba/roc/tests/hsapy/run_far_branch.py]: Remove print

File: numba/roc/tests/hsapy/test_compiler.py [hunk 5]
```diff
@@ -54,6 +54,7 @@ def tearDown(self):
         del self.cpu
 
 
+@unittest.skip("Test uses invalid BRIG based execution path.")
```
Comment (stuartarchibald) [inline numba/roc/tests/hsapy/test_compiler.py]: Just remove this?

File: MANIFEST.in [hunk 6]
```diff
@@ -8,3 +8,4 @@ recursive-include examples *.py
 prune docs/_build
 prune docs/gh-pages
 include numba/_version.py
+
```
Comment (sklam) [inline MANIFEST.in]: unnecessary changes.

Comment (stuartarchibald) [reply MANIFEST.in]: :) oh yes, thanks, that was from me trying to force in some data files that I changed my mind about, clearly left a space. Will fix.

**Comment by sklam** [comment]
how did `numba/caching.py` got in the diff with no change?

**Comment by stuartarchibald** [comment]
RE: https://github.com/numba/numba/pull/3023#issuecomment-403856241 its file mode is wrong. Moved 0x755->0x644

File: numba/roc/gcn_occupancy.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+wave_size = 64
+vector_register_file_size = 64 * 2**10  # 64 kB
+byte_per_VGPR = 4
+vgpr_per_simd = vector_register_file_size // byte_per_VGPR
+sgpr_per_simd = 512
+max_wave_count = 10
+max_inflight_wave_per_cu = max_wave_count * simd_per_cu
+
+# XXX due to limit in AMDGPU backend
+max_group_size = 256
```
Comment (sklam) [inline numba/roc/gcn_occupancy.py]: I recall there maybe changes to this.  Are there?

File: numba/roc/stubs.py [hunk 8]
```diff
@@ -99,6 +99,19 @@ def activelanepermute_wavewidth(src, laneid, identity, useidentity):
     raise _stub_error
 
 
+def ds_permute(src_lane, dest_lane):
```
Comment (sklam) [inline numba/roc/stubs.py]: What is "ds"?

Comment (stuartarchibald) [reply numba/roc/stubs.py]: From the [AMD GCN3 ISA Architecture Manual](http://32ipi028l5q82yhj72224m8j-wpengine.netdna-ssl.com/wp-content/uploads/2016/08/AMD_GCN3_Instruction_Set_Architecture_rev1.1.pdf) I think it's `Data Share`, see section 10.3.3 on `Data Share Indexed`.

File: numba/roc/gcn_occupancy.py (hunk 7 above)
Comment (stuartarchibald) [reply numba/roc/gcn_occupancy.py]: There are, but right now I'd elect to keep this as-is and update it as more information becomes available.

**Comment by stuartarchibald** [comment]
@sklam thanks for the review. I've made some changes based on things you noted and also things I picked up on self review. Am running this through the build farm now.

**Comment by stuartarchibald** [comment]
This is passing on the buildfarm on real AMDGCN hardware... merging.
